File: java/leet_code/88.java
/**
 * This class provides a solution for the problem of merging two sorted arrays.
 * The problem is solved by first making a copy of the first array, then iterating over the arrays
 * with two pointers, and copying the smaller element at each step back into the first array.
 * After that, if there are remaining elements in either of the arrays, they are copied back into
 * the first array.
 */
class Solution {

    /**
     * Merges two sorted arrays in-place into the first array.
     *
     * @param nums1 The first sorted array. It is an array of integers with a size of m + n, where m
     *              is the number of actual elements.
     * @param m     The number of actual elements in the first array. It is a non-negative integer.
     * @param nums2 The second sorted array. It is an array of integers with a size of n.
     * @param n     The number of actual elements in the second array. It is a non-negative integer.
     */
    public static void merge(final int[] nums1, final int m, final int[] nums2, final int n) {
        // Make a copy of the first array
        final int[] nums1_copy = new int[m];
        System.arraycopy(nums1, 0, nums1_copy, 0, m);

        // Initialize pointers for the first array, the second array, and the copy of the first array
        int p1 = 0;
        int p2 = 0;
        int p = 0;

        // Iterate over the arrays with the pointers
        while (p1 < m && p2 < n) {
            // If the current element in the copy of the first array is smaller than the current
            // element in the second array, copy it back into the first array and move the pointers
            if ((nums1_copy[p1] < nums2[p2])) {
                nums1[p++] = nums1_copy[p1++];
            } else {
                // Otherwise, copy the current element in the second array into the first array and
                // move the pointers
                nums1[p++] = nums2[p2];
                p2++;
            }
        }

        // If there are remaining elements in the copy of the first array, copy them back into the
        // first array
        if (p1 < m) {
            System.arraycopy(nums1_copy, p1, nums1, p1 + p2, m + n - p1 - p2);
        }

        // If there are remaining elements in the second array, copy them into the first array
        if (p2 < n) {
            System.arraycopy(nums2, p2, nums1, p1 + p2, m + n - p1 - p2);
        }
    }
}

File: scala/leet_code/88.scala
/** This module provides a solution to the problem of merging two sorted integer arrays into one sorted array. The first
  * array has a size that is large enough to hold the elements of the second array. The function `merge` is the main
  * function in this module.
  */
object Solution {

  /** Function to merge two sorted integer arrays into one sorted array. The first array has a size that is large enough
    * to hold the elements of the second array.
    *
    * @param nums1
    *   an Array[Int] representing the first sorted array. It has a size that is large enough to hold the elements of
    *   nums2.
    * @param m
    *   an Int representing the number of elements initially in nums1.
    * @param nums2
    *   an Array[Int] representing the second sorted array.
    * @param n
    *   an Int representing the number of elements in nums2.
    * @return
    *   Unit. The function modifies nums1 in-place to hold the merged array.
    */
  def merge(nums1: Array[Int], m: Int, nums2: Array[Int], n: Int): Unit = {
    // Initialize three pointers to keep track of the current index in nums1, nums2, and the merged array, respectively
    var i = m - 1
    var j = n - 1
    var k = m + n - 1

    // While there are elements in both nums1 and nums2
    while (i >= 0 && j >= 0) {
      // If the current element in nums1 is greater than the current element in nums2
      if (nums1(i) > nums2(j)) {
        // Place the current element in nums1 at the current index in the merged array
        nums1(k) = nums1(i)
        // Move to the previous element in nums1
        i -= 1
      } else {
        // Place the current element in nums2 at the current index in the merged array
        nums1(k) = nums2(j)
        // Move to the previous element in nums2
        j -= 1
      }
      // Move to the previous index in the merged array
      k -= 1
    }

    // If there are remaining elements in nums2
    while (j >= 0) {
      // Place the remaining elements in nums2 at the corresponding indices in the merged array
      nums1(k) = nums2(j)
      // Move to the previous element in nums2 and the previous index in the merged array
      j -= 1
      k -= 1
    }
  }
}
